aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java

```diff
@@ -87,18 +168,105 @@ public void close() throws IOException {
 
     super.close();
     closed = true;
+    currentStagingFile = null;
 
     try {
       stream.close();
 
-      s3.putObject(
-          PutObjectRequest.builder().bucket(location.bucket()).key(location.key()).build(),
-          RequestBody.fromFile(stagingFile));
+      completeUploads();
     } finally {
-      if (!stagingFile.delete()) {
-        LOG.warn("Could not delete temporary file: {}", stagingFile);
+      stagingFiles.forEach(f -> {
+        if (f.exists() && !f.delete()) {
+          LOG.warn("Could not delete temporary file: {}", f);
+        }
+      });
+    }
+  }
+
+  private void initializeMultiPartUpload() {
+    multipartUploadId = s3.createMultipartUpload(CreateMultipartUploadRequest.builder()
+        .bucket(location.bucket()).key(location.key()).build()).uploadId();
+  }
+
+  private void uploadParts() {
+    // exit if multipart has not been initiated
+    if (multipartUploadId == null) {
+      return;
+    }
+
+    stagingFiles.stream()
+        // do not upload the file currently being written
+        .filter(f -> currentStagingFile == null || !currentStagingFile.equals(f))
+        // do not upload any files that have already been processed
+        .filter(Predicates.not(multiPartMap::containsKey))
+        .forEach(f -> {
+          UploadPartRequest uploadRequest = UploadPartRequest.builder()
+              .bucket(location.bucket())
+              .key(location.key())
+              .uploadId(multipartUploadId)
+              .partNumber(stagingFiles.indexOf(f) + 1)
+              .contentLength(f.length())
+              .build();
+
+          CompletableFuture<CompletedPart> future = CompletableFuture.supplyAsync(
```

Comment:
If we want to make it async, can we directly leverage `S3AsyncClient`?

Reply:
That's a really good question.  Since it's the entire client that changes (right?) this would also require changes in the input stream.  Someone on the original thread commented they had problems with the async implementation, so I had steered away from it.

Reply:
Which original thread do you mean? Do you have a link?

Reply:
we use the StsAssumeRoleCredentialsProvider to keep a fresh session when running on prem. right now this doesnt' support StsAsyncClient and only supports StsClient.

Also the aws sdk provided nio client doesn't look like it supports proxies with basic auth (needed for us to leave our dc). The CRT one does, but it seems like it's still young?

https://github.com/aws/aws-sdk-java-v2/blob/master/services/sts/src/main/java/software/amazon/awssdk/services/sts/auth/StsAssumeRoleCredentialsProvider.java

Reply:
@jackye1995 the comment I was referring to about the async client was this: https://github.com/apache/iceberg/pull/1573#discussion_r502725465

Overall, I'm not sure we benefit much from the async client in this case because everything around is still synchronous.  I feel like if we move to a vectored-io implementation for read, that might be a better time to explore switching to async.

Reply:
Yeah I know there are some limitations right now for the NioNettyAsyncHttpClient. In this case let's just use the synchronous client then. Using `Tasks` suggested by Ryan sounds like a good idea.